microcosm_pubsub/producer.py

```diff
@@ -132,28 +170,30 @@ def generate_message_batches(self, messages, deferred_batch_size):
         for i in range(0, len(messages), deferred_batch_size):
             yield messages[i:i + deferred_batch_size]
 
-    def __exit__(self, type, value, traceback):
-        if type is not None:
-            return
-
-        messages = [
+    def construct_batch_pubsub_message(self, message_batch: List[PubsubMessage]):
+        return [
             dict(
-                media_type=media_type,
-                message=message,
-                topic_arn=topic_arn,
-                opaque_data=opaque_data,
+                media_type=pubsub_message.media_type,
+                message=pubsub_message.message,
+                message_attributes=pubsub_message.message_attributes,
+                opaque_data=pubsub_message.opaque_data,
+                topic_arn=pubsub_message.topic_arn,
             )
-            for media_type, message, topic_arn, opaque_data in self.messages
+            for pubsub_message in message_batch
         ]
 
-        for message_batch in self.generate_message_batches(messages, self.producer.deferred_batch_size):
+    def __exit__(self, type, value, traceback):
+        if type is not None:
+            return
+
+        for message_batch in self.generate_message_batches(self.messages, self.producer.deferred_batch_size):
             if len(message_batch) > 1:
                 self.producer.produce(
                     MessageBatchSchema.MEDIA_TYPE,
-                    messages=message_batch,
+                    messages=self.construct_batch_pubsub_message(message_batch)
```

Comment:
Had to do a bit of extra finagling here. Also need to verify that this actually works (I might break echo?)

Reply:
You're missing a trailing comma.